### PR TITLE
PUB-67 Added health check route

### PR DESCRIPTION
**Pull request by ishan123456789**
https://honesttech.atlassian.net/browse/PUB-67

### CODE REVIEW

File: src/service/pubsub.ts [hunk 1]
```diff
@@ -16,9 +17,52 @@ export default class PubSubService {
...
+      await PubSubService.client.getAllSubscriptionsOpenState();
+    const subscriptions = await PubSubService.getInstance().getSubscribers();
+    const notOpenSubs = subscriptions
+      .map(([_, metadata]) => metadata.subscriptionName)
+      .filter((subName) => {
+        const subState = subsOpenState.find(([name]) => name === subName);
+        return !subState?.[1];
+      });
+    if (notOpenSubs.length) {
+      Logger.Instance.warn(notOpenSubs, "These subs aren't open yet");
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:61]:
```suggestion
      Logger.Instance.warn({ subscriptions: notOpenSubs }, "These subs aren't open yet");
```

File: src/client/googlePubSub.ts [hunk 2]
```diff
@@ -525,4 +529,23 @@ export default class GooglePubSubAdapter implements PubSubClientV2 {
...
     return subscriptions.flat();
   }
+
+  public async getAllSubscriptionsOpenState(): Promise<IsOpenTuple[]> {
+    const subscriptions = await Bluebird.map(
+      Object.keys(this.projects),
+      async (project) => {
+        const [subscriptions] = await this.projects[
+          project
+        ].client.getSubscriptions();
```
Comment (rohit-gohri) [inline src/client/googlePubSub.ts:539]: We can use the cached subscriptions map, that way whole thing would sync (no need for Bluebird) and quicker
```suggestion
        const subscriptions = Array.from(
          this.projects[project].subscriptions.values(),
        );
```

File: src/client/googlePubSub.ts [hunk 3]
```diff
@@ -525,4 +529,23 @@ export default class GooglePubSubAdapter implements PubSubClientV2 {
...
+        return subscriptions.map(
+          ({ pubsub, metadata }) =>
+            [
+              metadata?.name?.split('/')?.slice(-1)[0] as string,
+              pubsub.isOpen as boolean,
+            ] as IsOpenTuple,
```
[the remark this answers is not in the document]
Comment (ishan123456789) [reply src/client/googlePubSub.ts:545]: Yes it throws some warnings

File: .changeset/many-bikes-cover.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@honestfoodcompany/pubsub': patch
```
Comment (rohit-gohri) [inline .changeset/many-bikes-cover.md:2]:
```suggestion
'@honestfoodcompany/pubsub': minor
```

File: src/client/googlePubSub.ts [hunk 5]
```diff
@@ -525,4 +529,20 @@ export default class GooglePubSubAdapter implements PubSubClientV2 {
     );
     return subscriptions.flat();
   }
+
+  public async getAllSubscriptionsOpenState(): Promise<IsOpenTuple[]> {
+    const subscriptions = await Bluebird.map(
+      Object.keys(this.projects),
+      async (project) => {
+        const subscriptions = await this.projects[
```
Comment (rohit-gohri) [inline src/client/googlePubSub.ts:537]: no need for await and bluebird

File: src/service/pubsub.ts [hunk 6]
```diff
@@ -16,9 +17,54 @@ export default class PubSubService {
...
+  }
+  public static async isHealthy(): Promise<boolean> {
+    if (PubSubService.status !== 'ready') {
+      Logger.Instance.warn('All subscriptions are not ready yet');
+      return false;
+    }
+
+    const subsOpenState =
+      await PubSubService.client.getAllSubscriptionsOpenState();
+    const subscriptions = await PubSubService.getInstance().getSubscribers();
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:52]: await not needed

File: src/service/pubsub.ts [hunk 7]
```diff
@@ -16,9 +17,54 @@ export default class PubSubService {
...
+          res.write(`is healthy`); //write a response to the client
+          res.end(); //end the response
+        } else {
+          res.statusCode = 500;
+          res.write(`Not healthy`); //write a response to the client
+          res.end(); //end the response
+        }
+      })
+      .listen(port); //the server object listens on port 8080
+    Logger.Instance.info(`Pubsub server running on port ${port}`);
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:42]: listen actually has a callback, log should go there

Comment (rohit-gohri) [inline src/service/pubsub.ts:42]: listen actually has a callback, log should go there

File: src/cli/subscriptions.ts [hunk 8]
```diff
@@ -27,4 +27,14 @@ require('yargs')
       describe: 'Path to exported credentials file',
       type: 'string',
     },
+
+    'pubsub-server-port': {
```
Comment (rohit-gohri) [inline src/cli/subscriptions.ts:31]: Makes sense to prefix env vars with PUBSUB but no need in cli args
```suggestion
    'server-port': {
```

File: src/service/pubsub.ts [hunk 9]
```diff
@@ -16,9 +17,51 @@ export default class PubSubService {
   private constructor() {
     this.initDriver();
     this.initClient();
+    this.startServer();
     this.bind(this);
   }
 
+  public startServer(): void {
+    if (process.env.RUN_HEALTH_SERVER !== 'true') return;
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:25]: Let's prefix this env var with PUBSUB too like all other?

```suggestion
    if (process.env.PUBSUB_HEALTH_SERVER !== 'true') return;
```

**Review by rohit-gohri — DISMISSED** [review]
Minor nitpicks about options/env vars naming. Rest LGTM

File: docs/getting-started/Installation.md [hunk 10]
```diff
@@ -56,11 +56,15 @@ GOOGLE_APPLICATION_CREDENTIALS=/path/to/gcp-project-83d5537a8388-key.json
...
+| `google-application-credentials` | `GOOGLE_APPLICATION_CREDENTIALS`  | see <https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account> to generate this                                                                                                                   |
+| `project-id`                     | `GOOGLE_CLOUD_PUB_SUB_PROJECT_ID` | name of the project in Google Cloud Platform                                                                                                                                                                                     |
+| `pubsub-server-port`             | `SERVER_PORT`                     | PORT at which the pubsub should run the server at                                                                                                                                                                                |
```
Comment (rohit-gohri) [inline docs/getting-started/Installation.md:64]:
```suggestion
| `server-port`             | `PUBSUB_SERVER_PORT`                     | PORT at which the pubsub should run the server at                                                                                                                                                                                |
```

File: src/service/pubsub.ts [hunk 11]
```diff
@@ -16,9 +17,51 @@ export default class PubSubService {
...
+      }); //the server object listens on port 8080
+  }
+  public static async isHealthy(): Promise<boolean> {
+    if (PubSubService.status !== 'ready') {
+      Logger.Instance.warn('All subscriptions are not ready yet');
+      return false;
+    }
+
+    const subsOpenState = PubSubService.client.getAllSubscriptionsOpenState();
+
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:52]: Match getSubscribers().length and subsOpenState.length

File: docs/getting-started/Installation.md [hunk 12]
```diff
@@ -56,11 +56,15 @@ GOOGLE_APPLICATION_CREDENTIALS=/path/to/gcp-project-83d5537a8388-key.json
...
+| `google-application-credentials` | `GOOGLE_APPLICATION_CREDENTIALS`  | see <https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account> to generate this                                                                                                                   |
+| `project-id`                     | `GOOGLE_CLOUD_PUB_SUB_PROJECT_ID` | name of the project in Google Cloud Platform                                                                                                                                                                                     |
+| `server-port`                    | `SERVER_PORT`                     | PORT at which the pubsub should run the server at                                                                                                                                                                                |
```
Comment (rohit-gohri) [inline docs/getting-started/Installation.md:64]:
```suggestion
| `server-port`                    | `PUBSUB_SERVER_PORT`                     | PORT at which the pubsub should run the server at                                                                                                                                                                                |
```

File: src/interface/pubSubClient.ts [hunk 13]
```diff
@@ -14,9 +14,10 @@ export default interface PubSubClient {
   ): Promise<string>;
   subscribe(subscriber: typeof Subscriber): void;
   getAllSubscriptions(): Promise<AllSubscriptions[]>;
+  getAllSubscriptionsOpen(): Promise<IsOpenTuple>;
```
Comment (rohit-gohri) [inline src/interface/pubSubClient.ts:17]: This is not needed I think
```suggestion
```

File: src/service/pubsub.ts [hunk 14]
```diff
@@ -16,9 +17,56 @@ export default class PubSubService {
...
+    }
+
+    const subsState = PubSubService.client.getAllSubscriptionsState();
+    const allSubs = PubSubService.getInstance().getSubscribers();
+
+    const notOpenSubs = subsState
+      .filter((subState) => !subState[1])
+      .map((subState) => subState[0]);
+    const allSubsNotHealthy = !!(
+      notOpenSubs.length && subsState.length === allSubs.length
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:59]: Subs are not healthy if either one if them is not open, or open subs count doesn't match total subs

```suggestion
      notOpenSubs.length || subsState.length !== allSubs.length
```

File: src/service/pubsub.ts [hunk 15]
```diff
@@ -16,9 +17,56 @@ export default class PubSubService {
...
+    if (subsState.length !== allSubs.length) {
+      Logger.Instance.warn("Some subs aren't active");
+    }
+    if (notOpenSubs.length) {
+      Logger.Instance.warn(
+        { subscriptions: notOpenSubs },
+        "These subs aren't open yet",
+      );
+    }
+    return !(subsState.length || subsState.length !== allSubs.length);
```
Comment (rohit-gohri) [inline src/service/pubsub.ts:67]:
```suggestion
    return !(notOpenSubs.length || subsState.length !== allSubs.length);
```

**Review by rohit-gohri — APPROVED** [review]

**Pull request merged**